machineconfiguration/v1/types.go

```diff
@@ -759,3 +760,123 @@ type ContainerRuntimeConfigList struct {
 
 	Items []ContainerRuntimeConfig `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigState describes the health of the Machines on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineConfigState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineConfigStateSpec `json:"spec"`
+	// +optional
+	Status MachineConfigStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineConfigStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineConfigState `json:"items"`
+}
+
+// MachineConfigStateSpec describes the type of State we are managing
+type MachineConfigStateSpec struct {
+	// configuration describes the object reference inforation of a MachineConfigState
+	Config MachineConfigStateConfig `json:"configuration"`
+}
+
+// MachineConfigStateStatus holds the reported information on a particular MachineConfigState
+type MachineConfigStateStatus struct {
+	// configuration Describes the Kind of MachineConfigState
+	Config MachineConfigStateConfig `json:"configuration"`
+	// mostRecentState is the most recent state reporting for each object in a particular machineState
+	// +listType=map
+	// +listMapKey=name
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	MostRecentState []ProgressionCondition `json:"mostRecentState" patchStrategy:"merge" patchMergeKey:"name"`
+	// progressionHistory contains a list of events that have happened on all objects in this machinestate
+	ProgressionHistory []ProgressionHistory `json:"progressionHistory"`
+	// mostRecentError is populated if the State reports an error.
+	MostRecentError string `json:"mostRecentError"`
+	// health reports the overall status of this MachineConfigState Given its Progress
+	Health MachineConfigStateHealthEnumeration `json:"health"`
```

Comment:
This in general indicates that we are still going down the route of using events. I was expecting that, when we get this underway, the MCD on each node would own directly updating their object, is that aligned?

I think if that's the case, we can be more detailed and specific with the information on the status here. I'd expect at the minimum a `Conditions`, a current version and desired version (desired should also be in spec but should be updated in status only when MCD has observed that it can start a migration to that version), probably an observedGeneration (so that we know whether MCD has seen the updated spec fields)

Is there anything missing there at the moment?

The different phases of draining, writing files etc initially would be represented as conditions, unless more specific information is needed that shouldn't be transient, WDYT? You know this area better than I do 😅

Reply:
So the way I have this built out is: 

1) the MCD Updates the node object's `machineconfiguration.openshift.io/reason: ` and `machineconfiguration.openshift.io/state: ` annotations which already exist so I can just augment them to be useful here.
2) the MachineStateController watches for node changes and then extrapolates what is going on from these annotations as well as: 
``` machineconfiguration.openshift.io/currentConfig: rendered-master-e0de67c47299c5c1814ed157a85a4f03
                    machineconfiguration.openshift.io/desiredConfig: rendered-master-e0de67c47299c5c1814ed157a85a4f03
                    machineconfiguration.openshift.io/desiredDrain: uncordon-rendered-master-e0de67c47299c5c1814ed157a85a4f03
                    machineconfiguration.openshift.io/lastAppliedDrain: uncordon-rendered-master-e0de67c47299c5c1814ed157a85a4f03
```
which all hold some useful info.

The MSC then updates the MachineConfigStates so that this logic lives in one place. The reason I am doing it this way is because when the operatorv1 work eventually lands I also want the MachineStateController to be updating those objects. We want the updating of state related objects to live in one place.

As for the Conditions I can swap out the `ProgressionHistory` for conditions but would definitely like to keep the `MostRecentState` for ease of view purposes.

ObservedGeneration is a good idea but for current/desired version, would you have any examples of how other operators manage that with objects such as this?

Reply:
having the MCD update the node rather than the machineconfigstate itself keeps api requests to a minimum since the MCD already does this in most places where I need information (I just had to make the information a bit more verbose and add a few extra calls). This also keeps in line for the design we have for the new controller.

Reply:
Discussed in slack, but, IMO introducing a middle man (new controller) adds complexity, increases the likelihood of a failure in the system and could lead to confused deputy style issues, vs just having the source of the information update the status of the new CRD. Though this is an implementation detail and doesn't necessarily affect the design of the API.